### DISCUSSION TITLE
Loads of crashes! After migrating to Axmol, now I'm very worried!

### DISCUSSION THREAD

**Discussion by meehjeevan** [axmolengine/axmol#2126] [Q&A]
I migrated my game from Cocos2d-x 3.17.2 and published an update on Google Play 1.5 months ago. I rolled it out to 5% of users for 3 weeks. A lot of crashes are showing up in the Play Console crash reports. I was already facing crash issues in the earlier version, and with Axmol, I encountered many crashes of new types. 
I’m really worried. I used the latest version from the dev channel on GitHub 2 months ago.
Is the release v2.1.5 stable?
How can I get rid of these kind of crashes? 

Please see below the an example of stack trace from play console mostly my app is getting :
I have attached full crash details in txt file also.
[crashes_stack_trace.txt](https://github.com/user-attachments/files/16842702/crashes_stack_trace.txt)

```
  #11  pc 0x0000000000750318  /vendor/lib64/egl/mt6789/libGLES_mali.so
  #12  pc 0x0000000000707120  /vendor/lib64/egl/mt6789/libGLES_mali.so (glTexImage2D+204)
  #13  pc 0x00000000006a4c2c  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (ax::backend::Texture2DGL::updateData(unsigned char*, unsigned long, unsigned long, unsigned long, int)+168) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #14  pc 0x00000000006865e4  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (ax::Texture2D::updateWithMipmaps(ax::_MipmapInfo*, int, ax::backend::PixelFormat, ax::backend::PixelFormat, int, int, bool, int)+608) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #15  pc 0x0000000000686f98  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (ax::Texture2D::updateWithImage(ax::Image*, ax::backend::PixelFormat, int)+912) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #16  pc 0x000000000068724c  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (ax::Texture2D::initWithImage(ax::Image*, ax::backend::PixelFormat)+116) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #17  pc 0x000000000068ce4c  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (ax::VolatileTextureMgr::reloadAllTextures()+112) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #18  pc 0x000000000059dc64  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/split_config.arm64_v8a.apk!libLudoX.so (Java_org_axmol_lib_AxmolRenderer_nativeInit+132) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
  #19  pc 0x0000000000098974  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/oat/arm64/base.odex (art_jni_trampoline+116)
  #20  pc 0x0000000000167578  /data/app/~~gOjnScfZZ26CyWbSijdBgw==/com.my.package-USJp4LiDBFqwwEBxGS0_Eg==/oat/arm64/base.odex (org.axmol.lib.AxmolRenderer.onSurfaceCreated+56)
  #21  pc 0x00000000005ba44c  /apex/com.android.art/lib64/libart.so (nterp_helper+9292)
  #22  pc 0x00000000004632ce  /system/framework/framework.jar (android.opengl.GLSurfaceView$GLThread.guardedRun+910)
  #23  pc 0x00000000005b8f54  /apex/com.android.art/lib64/libart.so (nterp_helper+3924)
  #24  pc 0x000000000046398c  /system/framework/framework.jar (android.opengl.GLSurfaceView$GLThread.run+52)
  #25  pc 0x000000000033a7a4  /apex/com.android.art/lib64/libart.so (art_quick_invoke_stub+612)
  #26  pc 0x000000000023a01c  /apex/com.android.art/lib64/libart.so (art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*)+144)
  #27  pc 0x0000000000539108  /apex/com.android.art/lib64/libart.so (art::Thread::CreateCallback(void*)+1600)
  #28  pc 0x0000000000538ab8  /apex/com.android.art/lib64/libart.so (art::Thread::CreateCallbackWithUffdGc(void*)+8)
  #29  pc 0x00000000000c17dc  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+204)
  #30  pc 0x0000000000054c30  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)
```

**Comment by rh101** [comment]
The call to `ax::VolatileTextureMgr::reloadAllTextures()` implies that this is recovering from a GL context loss, and it's attempting to reload all the textures.  By the look of the stack trace, the crash is happening in the OpenGL driver, and everything up to the point that it calls `glTexImage2D` in `Texture2DGL::updateData` seems to be correct.  So, for some reason, `glTexImage2D` is crashing

From the 3 phone models listed in the crash logs, only the `samsung a33x` shows what happens after the call to `glTexImage2D`:

```
samsung a33x (Galaxy A33 5G)

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 26624 >>> com.my.package <<<

backtrace:
  #00  pc 0x0000000001740c18  /vendor/lib64/egl/libGLES_mali.so (cobjp_convert_linear_to_block_16b_NxM+56)
  #01  pc 0x0000000001738dd8  /vendor/lib64/egl/libGLES_mali.so (cobjp_do_fastpath_reorder+8408)
  #02  pc 0x000000000173631c  /vendor/lib64/egl/libGLES_mali.so (cobjp_convert_pixels_fast+356)
  #03  pc 0x0000000001732afc  /vendor/lib64/egl/libGLES_mali.so (cobj_convert_pixels+316)
  #04  pc 0x00000000017353d8  /vendor/lib64/egl/libGLES_mali.so (cobj_convert_pixels_to_surface+552)
  #05  pc 0x00000000007500a4  /vendor/lib64/egl/libGLES_mali.so (gles_surface_convert_pixels_to_surface_internal+284)
  #06  pc 0x000000000074fa24  /vendor/lib64/egl/libGLES_mali.so (gles_surface_convert_pixels_to_surface+116)
  #07  pc 0x00000000007670c0  /vendor/lib64/egl/libGLES_mali.so (gles_texturep_upload_3d+1256)
  #08  pc 0x000000000076684c  /vendor/lib64/egl/libGLES_mali.so (gles_texturep_tex_image_3d_internal+1436)
  #09  pc 0x0000000000767728  /vendor/lib64/egl/libGLES_mali.so (gles_texturep_tex_image_2d_internal+72)
  #10  pc 0x000000000075996c  /vendor/lib64/egl/libGLES_mali.so (gles_texturep_tex_image_2d+388)
  #11  pc 0x000000000075962c  /vendor/lib64/egl/libGLES_mali.so (gles_texture_tex_image_2d+316)
  #12  pc 0x00000000007146d8  /vendor/lib64/egl/libGLES_mali.so (glTexImage2D+88)
  #13  pc 0x00000000006a4c2c  /data/app/~~uC3KHMy-G1HaUs7w1v2oJA==/com.my.package-JaPrs1y6HQbuvpkCpfMVww==/split_config.arm64_v8a.apk!libLudoX.so (ax::backend::Texture2DGL::updateData(unsigned char*, unsigned long, unsigned long, unsigned long, int)+168) (BuildId: f1f973bfee54dccee89664d5fde80229820a0bb7)
```

A similar issue is being tracked in #1211.  You can try a few things:

Option 1: Setting `AX_ENABLE_CACHE_TEXTURE_DATA` to 0, which automatically sets `AX_ENABLE_RESTART_APPLICATION_ON_CONTEXT_LOST` to 1 (only for Android).  What happens here is that if the GL context is no longer valid, the application will restart.  This it is seems to be handled in some other game engines.

Option 2: Handle it the way Cocos2d-x v4 does, which is by commenting out `super.onPause();` in `AxmolGLSurfaceView.java`, which will prevent the OS from triggering the context loss.  This seems to work, but there may be side-effects, although none of the developers using this have reported any issues:

```
    @Override
    public void onPause() {
        this.queueEvent(new Runnable() {
            @Override
            public void run() {
                AxmolGLSurfaceView.this.mRenderer.handleOnPause();
            }
        });
        //super.onPause(); //<<< comment this out
    }
```

That is until we can figure out why this is happening.

**Reply by meehjeevan** [reply]
so option 2 is the best to go for now ?

**Reply by rh101** [reply]
> so option 2 is the best to go for now ?

As I mentioned in my previous post, not calling `super.onPause();` may have side-effects.  More info [here](https://github.com/axmolengine/axmol/issues/1211#issuecomment-2136814428).

**Comment by rh101** [comment]
@halx99 Is there any chance the value for `GL_UNPACK_ALIGNMENT` set by `glPixelStorei` is incorrect when it is reloading the textures? If the alignment is not correct, then that is one possible reason `glTexImage2D` would crash:

```
void Texture2DGL::updateData(uint8_t* data, std::size_t width, std::size_t height, std::size_t level, int index)
{
    if (!_textureInfo.ensure(index, GL_TEXTURE_2D))
        return;

    // Set the row align only when mipmapsNum == 1 and the data is uncompressed
    auto mipmapEnabled = isMipmapEnabled(_textureInfo.minFilterGL) || isMipmapEnabled(_textureInfo.magFilterGL);
    if (!mipmapEnabled)
    {
        unsigned int bytesPerRow = width * _bitsPerPixel / 8;

        if (bytesPerRow % 8 == 0)
        {
            glPixelStorei(GL_UNPACK_ALIGNMENT, 8);
        }
        else if (bytesPerRow % 4 == 0)
        {
            glPixelStorei(GL_UNPACK_ALIGNMENT, 4);
        }
        else if (bytesPerRow % 2 == 0)
        {
            glPixelStorei(GL_UNPACK_ALIGNMENT, 2);
        }
        else
        {
            glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
        }
    }
    else
    {
        glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
    }

    glTexImage2D(GL_TEXTURE_2D, level, _textureInfo.internalFormat, width, height,
                 0,  // border
                 _textureInfo.format, _textureInfo.type, data);

    CHECK_GL_ERROR_DEBUG();

#if AX_ENABLE_CACHE_TEXTURE_DATA
    if (_generateMipmaps)
    {
        _hasMipmaps = false;
        generateMipmaps();
    }
#endif

    if (!_hasMipmaps && level > 0)
        _hasMipmaps = true;
}
```

**Comment by smilediver** [comment]
I don't know if this is related, but there's one unfixed issue in Axmol when app activity is recreated, which also destroys and recreates OpenGL context. You can repro it by starting your app, then change global font size in Android's settings menu, and go back to your app. This is the patch to fix it, you could try it:

```patch
From 220ff08dd46a3b2335b192a6b19997bf5673252c Mon Sep 17 00:00:00 2001
Date: Wed, 12 Jun 2024 19:19:49 +0300
Subject: [PATCH] Fix lost state in AxmolRenderer when app activity is
 relaunched

---
 .../java/src/org/axmol/lib/AxmolRenderer.java   | 17 +++++++++--------
 1 file changed, 9 insertions(+), 8 deletions(-)

diff --git a/core/platform/android/java/src/org/axmol/lib/AxmolRenderer.java b/core/platform/android/java/src/org/axmol/lib/AxmolRenderer.java
index 5b1c90030..f66677a08 100644
--- a/core/platform/android/java/src/org/axmol/lib/AxmolRenderer.java
+++ b/core/platform/android/java/src/org/axmol/lib/AxmolRenderer.java
@@ -46,8 +46,9 @@ public class AxmolRenderer implements GLSurfaceView.Renderer {
     private long mLastTickInNanoSeconds;
     private int mScreenWidth;
     private int mScreenHeight;
-    private boolean mNativeInitCompleted = false;
-    private boolean mIsPaused = false;
+
+    private static boolean gNativeInitCompleted = false;
+    private static boolean gIsPaused = false;
 
     // ===========================================================
     // Constructors
@@ -75,11 +76,11 @@ public class AxmolRenderer implements GLSurfaceView.Renderer {
         AxmolRenderer.nativeInit(this.mScreenWidth, this.mScreenHeight);
         this.mLastTickInNanoSeconds = System.nanoTime();
 
-        if (mNativeInitCompleted) {
+        if (gNativeInitCompleted) {
             // This must be from an OpenGL context loss
             nativeOnContextLost();
         } else {
-            mNativeInitCompleted = true;
+            gNativeInitCompleted = true;
         }
     }
 
@@ -161,17 +162,17 @@ public class AxmolRenderer implements GLSurfaceView.Renderer {
          * onSurfaceCreated is invoked. Can not invoke any
          * native method before onSurfaceCreated is invoked
          */
-        if (!mNativeInitCompleted)
+        if (!gNativeInitCompleted)
             return;
 
         AxmolRenderer.nativeOnPause();
-        mIsPaused = true;
+        gIsPaused = true;
     }
 
     public void handleOnResume() {
-        if (mIsPaused) {
+        if (gIsPaused) {
             AxmolRenderer.nativeOnResume();
-            mIsPaused = false;
+            gIsPaused = false;
         }
     }
 
-- 
```

**Comment by rh101** [comment]
Just an update.  I tried to test this issue, and managed to reproduce it very easily.  Firstly, set the following to `false` in `AxmolActivity.java`:

`this.mGLSurfaceView.setPreserveEGLContextOnPause(false);`

In this specific instance, what crashed the app consistently on recovery from the context loss is a 999x54 sized texture which is added in `Director::createStatsLabel()`:

```
void Director::createStatsLabel()
{
...
    unsigned char* data = nullptr;
    ssize_t dataLength  = 0;
    getFPSImageData(&data, &dataLength);

    Image* image = new Image();
    bool isOK    = image->initWithImageData(data, dataLength, false);
    if (!isOK)
    {
        if (image)
            delete image;
        AXLOGE("{}", "Fails: init fps_images");
        return;
    }

    texture = _textureCache->addImage(image, "/ax_fps_images", PixelFormat::RGBA4);
    AX_SAFE_RELEASE(image);
...
```

The crash always happened with this call in `Texture2DGL::updateData`:
```
glTexImage2D(GL_TEXTURE_2D, level, _textureInfo.internalFormat, width, height,
                 0,  // border
                 _textureInfo.format, _textureInfo.type, data);
```

This was due to the `GL_UNPACK_ALIGNMENT` being set to 8 just prior to the call to `glTexImage2D`.  As a test, I changed it so it calls it with 1, `glPixelStorei(GL_UNPACK_ALIGNMENT, 1)`, and the crash never happened.  So, the data being sent to `Texture2DGL::updateData` is not correct.

Change the pixel format used in the call to `addImage` in `Director::createStatsLabel()` fixed the problem.

From:
`texture = _textureCache->addImage(image, "/ax_fps_images", PixelFormat::RGBA4);`

To one of the following:
`texture = _textureCache->addImage(image, "/ax_fps_images", image->getPixelFormat());`
or
`texture = _textureCache->addImage(image, "/ax_fps_images", PixelFormat::RGBA8);`

The crash never happened after that.

Tracing it through, the correct `GL_UNPACK_ALIGNMENT` ends up as 4 when the correct pixel format is used.  With the incorrect pixel format, the `GL_UNPACK_ALIGNMENT` ends up being 8, which causes the crash.

Note that *ALL* resources that exist on disk were loaded correctly, since their pixel format is always set to the correct value.  At no point could I get it to crash on reloading over 500MB of textures from disk after fixing the FPS texture pixel format.

@meehjeevan I'm not sure if this is what is causing the crash that you are seeing, but if you want to test it out, set `AX_STRIP_FPS` to 1 in the config, which you can do in your project `CMakeLists.txt`:

```
if(ANDROID)
    add_definitions(-DAX_STRIP_FPS=1)
endif()
```

This will no longer create the texture that is causing the problem. If this doesn't fix the issue for you, then you can test your own app to find exactly which cached texture is causing the problem by setting this `this.mGLSurfaceView.setPreserveEGLContextOnPause(false);`.  Either follow the call stack backwards after the crash, or put a breakpoint on `void VolatileTextureMgr::reloadAllTextures()` and trace it from there.

PR #2138 has the fix for this specific issue.

UPDATE: This issue had nothing to do with the FPS texture pixel format, but rather something else down the line that was affecting it. This has been resolved in the PR linked above.

**Reply by meehjeevan** [reply]
`this.mGLSurfaceView.setPreserveEGLContextOnPause(false);
`
I wasn't able to reproduce the crash with this change. Does it work only when stats is showing?

So, according to your update: I don't have to make changes in `CMakeLists.txt`?

**Reply by rh101** [reply]
No changes to your project are required once that PR is merged.

**Reply by meehjeevan** [reply]
So, until the PR is merged, which option should I choose? While using version Release-2.1.5. 
1.  commenting out super.onPause(); in AxmolGLSurfaceView.java
2. add_definitions(-DAX_STRIP_FPS=1) in CMakeLists.txt

**Reply by rh101** [reply]
It is now merged in.  If for some reason you need to keep using 2.1.5, then simply back-port the changes to your own branch of the engine code.

**Reply by meehjeevan** [reply]
ok thanks, there's no specific reason for using 2.1.5 but just for stability as it's release version. Please suggest should I choose latest dev channel or release 2.1.5 for it to be more stable ?

**Reply by rh101** [reply]
> Please suggest should I choose latest dev channel or release 2.1.5 for it to be more stable ?

That is entirely up to you.  If you use the `DrawNode`, then perhaps wait till 2.2 is released, since there may be an issue with the new `DrawNode` implementation the dev branch.

It's straightforward to merge the required changes from PR #2138 into your 2.1.5. The only changes you need are these:
![image](https://github.com/user-attachments/assets/efdb1599-5deb-4c93-9c72-c7b36850d9cc)

**Reply by meehjeevan** [reply]
I tried to merge the required changes from PR https://github.com/axmolengine/axmol/pull/2138 into your 2.1.5 but there was a lot of conflicts so I didn't play with that all. So, I tried to build with latest from dev branch and then I got this error:
`proj.android/app/.cxx/Debug/6a2a4q1u/arm64-v8a/android_gradle_build.json debug|arm64-v8a : expected buildFiles file '/axmol/extensions/DrawNodeEx/CMakeLists.txt' to exist
`
Any idea or ETA when 2.2.0 will be available?

**Reply by rh101** [reply]
You need to do a `clean`, and if that doesn't work, then delete the build folder, and compile it again.  There is no build issue with the current dev branch.

Also, in my previous post when I mentioned "merge the required changes", I quite literally meant the changes to the two functions, `TextureBackend::updateTextureDescriptor` and `Texture2DGL::initWithZeros()`.  There should be no reason at all to have any kind of conflicts when building with those changes.
